### ISSUE TITLE
DTLS attibutes are stripped for non-RTP protocols

### ISSUE THREAD

**Issue by abogosyan** [sipwise/rtpengine#971]
Hello, it looks like DTLS attributes are stripped for unsupported protocols, for example, somewhere in WebRTC offer: 
```
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=ice-ufrag:u6F9
a=ice-pwd:Al47qnr3XIeupDeBmP5vbMSF
a=ice-options:trickle
a=fingerprint:sha-256 F3:56:DA:3F:0B:B6:DD:88:A6:01:64:3C:A1:B4:95:03:DD:9D:32:CB:F1:A0:1D:3E:04:E0:07:46:E2:AD:95:49
a=setup:actpass
a=mid:data
a=sctp-port:5000
a=max-message-size:262144
```

And processed offer is:
```
m=application 30060 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 127.0.0.1
a=sctp-port:5000
a=max-message-size:262144a=mid:data
a=sendrecv
```
I believe that  DTLS should remain untouched, ICE should be processed, isn't it?

**Comment by rfuchs** [comment]
Hm ok, I assume you're just passing the DTLS stuff through?

**Comment by abogosyan** [comment]
Nope, it's webrtc case, I want to decrypt savpf -> avp and have NAT traversal (ice), but don't want to touch sctp stream security because rtpengine wouldn't decrypt it

**Comment by rfuchs** [comment]
Hm ok, seeing as SCTP is involved, I guess the connection details (port and `c=` line) also need to remain untouched.

**Comment by rfuchs** [comment]
002e620 is a very primitive patch. Is this sufficient for you?

**Comment by rfuchs** [comment]
Which `ICE=...` mode are you using?

**Comment by abogosyan** [comment]
> Which `ICE=...` mode are you using?

I use ICE=remove to process webrtc offer and ICE=force in the opposite direction

**Comment by rfuchs** [comment]
Ok, try with 5b8f74dd then.

DTLS passthrough should be fine like this, as rtpengine only looks for and processes DTLS if the protocol is known to be DTLS-enabled.

**Comment by abogosyan** [comment]
> Ok, try with [5b8f74d](https://github.com/sipwise/rtpengine/commit/5b8f74dd4946f5aa10715efa196617994c168cba) then.
> 
> DTLS passthrough should be fine like this, as rtpengine only looks for and processes DTLS if the protocol is known to be DTLS-enabled.

Tried new patch, looks like SDP part is OK now. We are in process of testing ICE and DTLS passthrough for SCTP stream.

**Comment by abogosyan** [comment]
Looks like rtpengine still doing DTLS handshake by himself instead of passthrough it for unknown transports

**Comment by abogosyan** [comment]
Tried some fixed:
- Added `other->protocol` check here for other media too and clear DTLS flags if it's unknown https://github.com/sipwise/rtpengine/blob/master/daemon/call.c#L1295
- that resulted in rtpengine stopped to process DTLS packets by himself, but still doesn't forward it :(

Also found the place where DTLS flag can also be set: https://github.com/sipwise/rtpengine/blob/master/daemon/call.c#L1716
added here check for other->protocol but it didn't help

**Comment by abogosyan** [comment]
Also found that ICE isn't concluded for unknown protocol, because it's expecting two components (like for RTP transports).

**Comment by rfuchs** [comment]
I've included some of your suggested fixes in the branch, we're now at 89182ab7

The change in `__dtls_logic` shouldn't be necessary as it's overridden by `__generate_crypto`

I'm not sure why DTLS packets shouldn't be forwarded, because the only place DTLS is demuxed from RTP is in `media_demux_protocols` and it depends on the presence of the `DTLS` flag. So if that isn't set, then DTLS should be treated like any other packets.

Unfortunately I don't have a test bed for this use case, so I can't try to reproduce it.

**Comment by abogosyan** [comment]
> I've included some of your suggested fixes in the branch, we're now at [89182ab](https://github.com/sipwise/rtpengine/commit/89182ab79b405aac8bcef5581abf2da5bd2d40f7)
> 
> The change in `__dtls_logic` shouldn't be necessary as it's overridden by `__generate_crypto`
> 
> I'm not sure why DTLS packets shouldn't be forwarded, because the only place DTLS is demuxed from RTP (or other packets) is in `media_demux_protocols` and it depends on the presence of the `DTLS` flag. So if that isn't set, then DTLS should be treated like any other packets.
> 
> Unfortunately I don't have a test bed for this use case, so I can't try to reproduce it.

It wasn't forwarded because ICE wasn't concluded for unknown media type (expected 2 components but got 1)
